src/game/movement.rs

```diff
@@ -120,36 +119,118 @@ fn update_facing(mut player_query: Query<(&MovementController, &mut Sprite)>) {
     }
 }
 
-/// Time between walk sound effects.
+/// Update the animation.
+fn update_animation(
+    time: Res<Time>,
+    mut query: Query<(&mut PlayerAnimation, &MovementController, &mut TextureAtlas)>,
+) {
+    for (mut animation, controller, mut atlas) in &mut query {
+        animation.update_timer(time.delta());
+
+        let animation_state = if controller.0 == Vec2::ZERO {
+            PlayerAnimationState::Idle
+        } else {
+            PlayerAnimationState::Walking
+        };
+        animation.update_state(animation_state);
+
+        if animation.changed() {
+            atlas.index = animation.get_atlas_index();
+        }
+    }
+}
+
+/// If the player is moving, play a step sound effect synchronized with the animation.
+fn trigger_step_sfx(mut commands: Commands, mut step_query: Query<&PlayerAnimation>) {
+    for animation in &mut step_query {
+        if animation.state == PlayerAnimationState::Walking
+            && animation.changed()
+            && (animation.frame == 2 || animation.frame == 5)
+        {
+            commands.trigger(Sfx::Step);
+        }
+    }
+}
+
+/// Component that tracks player's animation state.
+/// It is tightly bound to the texture atlas we use.
 #[derive(Component, Reflect)]
 #[reflect(Component)]
-pub struct StepSfx {
-    pub cooldown_timer: Timer,
+pub struct PlayerAnimation {
+    timer: Timer,
+    frame: usize,
+    state: PlayerAnimationState,
 }
 
-impl StepSfx {
-    pub fn new(cooldown: Duration) -> Self {
-        let mut cooldown_timer = Timer::new(cooldown, TimerMode::Once);
-        cooldown_timer.set_elapsed(cooldown);
-        Self { cooldown_timer }
-    }
+#[derive(Reflect, PartialEq)]
+pub enum PlayerAnimationState {
+    Idle,
+    Walking,
 }
 
-fn tick_step_sfx(time: Res<Time>, mut step_query: Query<&mut StepSfx>) {
-    for mut step in &mut step_query {
-        step.cooldown_timer.tick(time.delta());
+impl PlayerAnimation {
```

Comment:
Could we move the animation stuff to a new file?

Reply:
Done